### ISSUE TITLE
GAMATE input broken

### ISSUE THREAD

**Issue by Fabrizio-Caruso** [cc65/cc65#515]
GAMATE joystick input seems to be broken (it always register UP).

The code I have used to test this behavior is own my game CROSS CHASE. The very same code is successfully used on all other joystick/gamepad-equipped CC65-targets:
```
	#include <joystick.h>
	void movePlayerByJoystick(unsigned char joyInput)
	{
		if(JOY_UP(joyInput))
		{
			_DO_MOVE_UP
		}
		else if(JOY_DOWN(joyInput))
		{
			_DO_MOVE_DOWN
		}
		else if(JOY_LEFT(joyInput))
		{
			_DO_MOVE_LEFT
		}
		else if(JOY_RIGHT(joyInput))
		{
			_DO_MOVE_RIGHT
		}
		else if(JOY_BTN_1(joyInput) && guns>0 && !missile._status)
		{
			playerFire = 1;
		}
			_DRAW_PLAYER	
	}	
```
where the parameter is passed the value 
`joy_read(JOY_1)`

**Comment by mrdudz** [comment]
Am Dienstag, 7. November 2017, 23:01:58 CET schrieb Fabrizio Caruso:
> Hi
> 
> GAMATE joystick input is broken (it always register UP).
> 
> The code I have used it to test is my game, which is the same which I use on
> all other joystick/gamepad-equipped CC65-targets: ```
> 	#include <joystick.h>
> 	void movePlayerByJoystick(unsigned char joyInput)
> 	{
> 		if(JOY_UP(joyInput))
> 		{
> 			_DO_MOVE_UP
> 		}
> 		else if(JOY_DOWN(joyInput))
> 		{
> 			_DO_MOVE_DOWN
> 		}
> 		else if(JOY_LEFT(joyInput))
> 		{
> 			_DO_MOVE_LEFT
> 		}
> 		else if(JOY_RIGHT(joyInput))
> 		{
> 			_DO_MOVE_RIGHT
> 		}
> 		else if(JOY_BTN_1(joyInput) && guns>0 && !missile._status)
> 		{
> 			playerFire = 1;
> 		}
> 			_DRAW_PLAYER
> 	}
> ```
> where the parameter is passed the value
> `joy_read(JOY_1)`

pretty sure i tested that it works, using the conio example most likely. i 
didnt use those ugly macros (which should die in a fire IMHO) though, so 
perhaps that is what is broken...

-- 

http://www.hitmen-console.org    http://magicdisk.untergrund.net
http://www.pokefinder.org        http://ar.pokefinder.org

I don't want to achieve immortality through my work... I want to achieve it 
through not dying. 
<Woody Allen>

**Comment by Fabrizio-Caruso** [comment]
There was some re-factoring of the joystick code/macros not too long ago that broke the Vic 20 target badly (crash when using the joystick driver). 
Here the problem is different: the code or just the macro always registers UP.
Maybe the refactoring broke something.

**Comment by mrdudz** [comment]
Am Mittwoch, 8. November 2017, 12:07:48 CET schrieb Fabrizio Caruso:
> There was some re-factoring of the joystick code/macros not too long ago
> that broke the Vic 20 target badly (crash when using the joystick driver).
> Here the problem is different: the code or just the macro always registers
> UP. Maybe the refactoring broke something.

OK, perhaps. unfortunately i am busy with other stuff atm, someone else would 
have to look at it. it shouldnt be terribly hard to fix, i guess :)

-- 

http://www.hitmen-console.org    http://magicdisk.untergrund.net
http://www.pokefinder.org        http://ar.pokefinder.org

Ich bin nicht faul.... ich hab nur keine Lust.

**Comment by Fabrizio-Caruso** [comment]
Your previous quote "Ich bin nicht faul.... ich hab nur keine Lust.". was it random? LOL

Anyway you are probably right about the macros! The conio.c test seems to detect the different movements and buttons.

So the most likely culprits are the macros.

**Comment by mrdudz** [comment]
Am Mittwoch, 8. November 2017, 17:12:10 CET schrieb Fabrizio Caruso:
> Your previous quote "Ich bin nicht faul.... ich hab nur keine Lust.". was it
> random? LOL

haha, yes, of course =)

> 
> Anyway you are probably right about the macros! The conio.c test seems to
> detect the different movements and buttons.
> 
> So the most likely culprits are the macros.

should be easy to fix then, i guess :)

-- 

http://www.hitmen-console.org    http://magicdisk.untergrund.net
http://www.pokefinder.org        http://ar.pokefinder.org

Motivation ist der Anfang allen Übels

**Comment by Fabrizio-Caruso** [comment]
The mask needs to be inverted. If I pass the bit-wise not of the joystick input, I get the correct behavior!!

**Comment by Fabrizio-Caruso** [comment]
In include/gamate.h, changing the masks with their bit-wise NOT value may fix this issue, i.e.,
-#define JOY_UP_MASK     0x01
-#define JOY_DOWN_MASK   0x02
-#define JOY_LEFT_MASK   0x04
-#define JOY_RIGHT_MASK  0x08
-#define JOY_BTN_1_MASK  0x10
-#define JOY_BTN_2_MASK  0x20
-#define JOY_BTN_3_MASK  0x40
-#define JOY_BTN_4_MASK  0x80
+#define JOY_UP_MASK     0xFE
+#define JOY_DOWN_MASK   0xFD
+#define JOY_LEFT_MASK   0xFB
+#define JOY_RIGHT_MASK  0xF7
+#define JOY_BTN_1_MASK  0xEF
+#define JOY_BTN_2_MASK  0xDF
+#define JOY_BTN_3_MASK  0xBF
+#define JOY_BTN_4_MASK  0x7F

**Comment by Fabrizio-Caruso** [comment]
I have created the pull request "Fix joy mask in gamate #516", which ONLY changes the bit masks.
This pull request is ONLY TENTATIVE. It still needs some serious testing.

**Comment by oliverschmidt** [comment]
@mrdudz:

> i didnt use those ugly macros (which should die in a fire IMHO) though, so perhaps that is what is broken...

Currently it seems that - totally independent from my refactoring - the GAMATE joy_read is absolutely incompatible to the JOY_xxx macros. Given your statement above I'm now wondering if the GAMATE joy_read indeed was just never compatible with them - as you didn't care about them. Is that a plausible scenario?

Let me state that I don't consider it appropriate to sabotage the cc65 cross target development approach by deliberately implementing functions in a way incompatible with all other targets.

**Comment by mrdudz** [comment]
Am Donnerstag, 9. November 2017, 14:05:54 CET schrieb Oliver Schmidt:
> @mrdudz:
> > i didnt use those ugly macros (which should die in a fire IMHO) though, so
> > perhaps that is what is broken...
> Currently it seems that - totally independent from my refactoring - the
> GAMATE joy_read is absolutely incompatible to the JOY_xxx macros. Given
> your statement above I'm now wondering if the GAMATE joy_read indeed was
> just never compatible with them - as you didn't care about them. Is that a
> plausible scenario?

sure, as said, i didnt ever use them :) until the refactoring recently i didnt 
even know they existed =P

> Let me state that I don't consider it appropriate to sabotage the cc65 cross
> target development approach by deliberately implementing functions in a way
> incompatible with all other targets.

certainly wasnt deliberately :) (and i totally agree)

-- 

http://www.hitmen-console.org    http://magicdisk.untergrund.net
http://www.pokefinder.org        http://ar.pokefinder.org

Wenn in Deutschland der Karneval den Karnevalisten gehoert, wer betreibt in 
Oesterreich dann den Fasching?

**Comment by oliverschmidt** [comment]
@mrdudz:

>> Is that a plausible scenario?

> sure [...]

Thanks for the feedback! Then I think it's safe to merge a change adding the EOR $FF.

> certainly wasnt deliberately :) (and i totally agree)

Thanks for clarifying that :-)

**Comment by Fabrizio-Caruso** [comment]
Thanks everyone! 
The issue has now been solved by my PR #517
